### ISSUE TITLE
SPOE: When using balance source backend agent/server selection silently fails

### ISSUE THREAD

**Issue by daniel-corbett** [haproxy/haproxy#1797]
### Detailed Description of the Problem

Hi,

Encountered an issue with `balance source` and an SPOE agent backend that causes HAProxy to not make any connections to the agent. It seems to silently fail in the background as it appears to not be able to select an agent/server.

I realize that `balance source` may not be ideal in this situation but this bug can be triggered if a user has specified it within their defaults section.

### Expected Behavior

HAProxy should either select an agent/server for processing of the request, present a configuration error indicating that `balance source` is not supported in an SPOA backend, or present an error when this failure occurs.

### Steps to Reproduce the Behavior

1. Create an HAProxy configuration that uses either `balance source` in the defaults section or SPOA backend
2. Start haproxy with `strace -e connect -f ./haproxy -f haproxy.cfg -d`
3. Send request through listen/frontend that should trigger connection to SPOA
4. Observe that no connect() functions are called

### Do you have any idea what may have caused this?

Seems to be that the SPOE connection is not setting a valid `src->ss_family` for the stream/connection.

https://github.com/haproxy/haproxy/blob/ca6799297990f0f17d81a5e0fd30d116b6c4455f/src/backend.c#L747-L750

### What is your configuration?

```haproxy
haproxy.cfg

defaults
  balance source

listen fe_main
  mode http
  bind 127.0.0.1:8282
  filter spoe engine spoe config spoe.conf
  http-request return status 200

backend be_spoe
  mode tcp
  server spoa1 unix@/var/run/spoa.sock
```

spoe.conf
```
[spoe]
    spoe-agent spoa
    messages test
    timeout hello 1s
    timeout idle 1s
    timeout processing 1s
    option var-prefix test
    use-backend be_spoe

spoe-message test
    args src
    event on-client-session
```
```


### Output of `haproxy -vv`

```plain
HAProxy version 2.6.1-f6ca66d 2022/06/21 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2027.
Known bugs: http://www.haproxy.org/bugs/bugs-2.6.1.html
Running on: Linux 5.4.0-122-generic #138-Ubuntu SMP Wed Jun 22 15:00:31 UTC 2022 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_PCRE=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1 USE_SYSTEMD=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : +EPOLL -KQUEUE +NETFILTER +PCRE -PCRE_JIT -PCRE2 -PCRE2_JIT +POLL +THREAD +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -ENGINE +GETADDRINFO +OPENSSL +LUA +ACCEPT4 -CLOSEFROM +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL -PROCCTL +THREAD_DUMP -EVPORTS -OT -QUIC -PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=8).
Built with OpenSSL version : OpenSSL 1.1.1f  31 Mar 2020
Running on OpenSSL version : OpenSSL 1.1.1f  31 Mar 2020
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with network namespace support.
Support for malloc_trim() is enabled.
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE version : 8.39 2016-06-14
Running on PCRE version : 8.39 2016-06-14
PCRE library supports JIT : no (USE_PCRE_JIT not set)
Encrypted password support via crypt(3): yes
Built with gcc compiler version 9.4.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : none

Available filters :
	[CACHE] cache
	[COMP] compression
	[FCGI] fcgi-app
	[SPOE] spoe
	[TRACE] trace
```

**Comment by capflam** [comment]
Indeed, the same happens if you use a unix socket on the bind line. For the SPOE, the client is an applet, there is no source address. For a client connected to a unix socket, there is no IP address. In both case, a internal error is reported. However, for other algorithms, there is a fallback on round robin if no server is assigned.  The same should be performed in this case. So the `else` statement should be removed.

I will push a fix to do so. It will probably be shipped with the next 2.6 release. Thanks !

**Comment by capflam** [comment]
It should be fixed now. It will be shipped with the 2.6.2.

**Comment by daniel-corbett** [comment]
Thanks Christopher! Have a great day.
